video/out/vulkan/context_display.c

```diff
@@ -0,0 +1,461 @@
+        }
+
+        ret = true;
+    }
+    mp_dbg(log, "Parsed Display Spec: %d, %d, %d\n",
+           parsed_display, parsed_mode, parsed_plane);
+
+
+  finish:
+    if (display) {
+        *display = parsed_display;
+    }
+    if (mode) {
+        *mode = parsed_mode;
+    }
+    if (plane) {
+        *plane = parsed_plane;
+    }
+    return ret;
+}
+
+static bool print_display_info(struct mp_log *log) {
+    bool ret = false;
+    VkResult res;
+    VkInstance inst;
+    VkPhysicalDevice *devices = NULL;
+    VkDisplayPropertiesKHR *props = NULL;
+    VkDisplayModePropertiesKHR *modes = NULL;
+    VkDisplayPlanePropertiesKHR *planes = NULL;
+    VkDisplayKHR **planes_to_displays = NULL;
+
+    // Create a dummy instance to list the resources
+    VkInstanceCreateInfo info = {
+        .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
+        .enabledExtensionCount = 1,
+        .ppEnabledExtensionNames = (const char*[]) {
+            VK_KHR_DISPLAY_EXTENSION_NAME
+        },
+    };
+
+    res = vkCreateInstance(&info, NULL, &inst);
+    if (res != VK_SUCCESS)
+        goto done;
+
+    uint32_t num_devices = 0;
+    res = vkEnumeratePhysicalDevices(inst, &num_devices, NULL);
+    if (res != VK_SUCCESS)
+        goto done;
+
+    if (!num_devices) {
+        mp_info(log, "No Vulkan devices detected.\n");
+        ret = true;
+        goto done;
+    }
+
+    devices = talloc_array(NULL, VkPhysicalDevice, num_devices);
+    vkEnumeratePhysicalDevices(inst, &num_devices, devices);
+    if (res != VK_SUCCESS)
+        goto done;
+
+    mp_info(log, "Vulkan Devices:\n");
+    for (int i = 0; i < num_devices; i++) {
+        VkPhysicalDevice device = devices[i];
+
+        VkPhysicalDeviceProperties prop;
+        vkGetPhysicalDeviceProperties(device, &prop);
+        mp_info(log, "  '%s' (GPU %d, ID %x:%x)\n", prop.deviceName, i,
+                (unsigned)prop.vendorID, (unsigned)prop.deviceID);
+
+        // Count displays. This must be done before enumerating planes with the
+        // Intel driver, or it will not enumerate any planes. WTF.
+        int num_displays;
+        res = vkGetPhysicalDeviceDisplayPropertiesKHR(device, &num_displays, NULL);
+        if (res != VK_SUCCESS) {
+            mp_warn(log, "    Failed counting display properties\n");
+            continue;
+        }
+
+        if (!num_displays) {
+            mp_info(log, "    No available displays for device.\n");
+            continue;
+        }
+
+        // Enumerate Planes
+        int num_planes;
+        res = vkGetPhysicalDeviceDisplayPlanePropertiesKHR(device, &num_planes, NULL);
+        if (res != VK_SUCCESS) {
+            mp_warn(log, "    Failed counting planes\n");
+            continue;
+        }
+
+        if (!num_planes) {
+            mp_info(log, "    No available planes for device.\n");
+            continue;
+        }
+
+        planes = talloc_array(NULL, VkDisplayPlanePropertiesKHR, num_planes);
```

Comment:
This function leaks a lot of memory, especially in the failure cases. It's also confusing to track the lifetime of all those allocations. I suggest grouping all these allocations under a single `tmp` object that you then free at the end the function.